feat: create*WireAdapterMock instead of register*

jodarove opened the pull request:
## Details
Components that use wire adapters to get data shouldn't depend on outside code or data, thus, you usually mock modules exposing wire adapters and utilize this library to provide the data, doing so your test does not depend on unpredictable factors like remote invocation or server latency.

With the [Wire protocol reform](https://github.com/salesforce/lwc-rfcs/blob/master/text/0000-wire-reform.md), the wire-service now requires a wire-adapter (even a mock) in the proper format; otherwise, the test to a component using a wire will throw during rendering. When [upgrading the lwc-test repository to work with lwc 1.5.0](https://github.com/salesforce/lwc-test/pull/76), we had to hijack the decorator's registration system of the engine to replace the used wire-adapter (maybe invalid) with a valid one, thus providing backward-compatibility for existing unit tests.

This PR solves #15 and attempts to start the deprecation/removal of such mechanism by:
1. Adding a new function (`createWireAdapterMock`) to the library API, `createWireAdapterMock()` returns a fully functional wire adapter mock that works with the `@wire` of a component.
2. The adapter mock returned from `createWireAdapterMock()` has the `emit()` and other necessary APIs exposed to tests.
3. The `register*TestWireAdapter` call is no longer necessary for wire adapter mocks returned from createWireAdapterMock()`. 
4. For backwards compatibility, the `register*TestWireAdapter` calls will do the magic of converting a standard `jest.fn()` to a `createWireAdapterMock()` mock.
5. We deprecate the `register*TestWireAdapter` call, doing our due diligence of outputting a warning when used. We should also update documentation,  sample repositories, and change the default base component stubs.

Open questions:
- What about custom written APEX wire adapters? More context in [this example](https://trailhead.salesforce.com/content/learn/modules/test-lightning-web-components/write-a-jest-test-for-wire-service). 
Possible solutions:
1. Customers will have to mock them.
2. All `@salesforce/apex/($1).($2)` will be automatically mocked in the lightning stubs. (platform case)
3. Modify [@lwc/jest-transform](https://github.com/salesforce/lwc-test/blob/master/packages/%40lwc/jest-transformer/src/transforms/apex-scoped-import.js) and return a fully functional mock adapter. (including same api to that one returned by `createWireAdapterMock()`)

pmdartus reviewed (commented):
As we are redesigning this API, I think it's a good time to address this specific issue https://github.com/salesforce/wire-service-jest-util/issues/15. IMO, the fact that you can't target a specific adapter a huge flaw with this library. I don't think this proposal solves this specific issue.

ekashida commented on src/index.js:
```diff
@@ -4,35 +4,118 @@
+    return identifier;
+}
+
+function createWireAdapterMock(fn, adapterObserver) {
```
I didn't see `fn` being passed in for any of the tests, what is the use case for it as an argument?

jodarove replied:
it's only used in the [apex test](https://github.com/salesforce/wire-service-jest-util/pull/36/files#diff-603feae20112d93be4a6a77e3b6636f9982b9c95475ab1b9af1711ae20fc5935R9)

yungcheng commented on README.md:
````diff
@@ -17,18 +17,21 @@ export default class MyComponent extends LightningElement {
 
+Mock the `getTodo` wire adapter.
+
+```js
+import { createTestWireAdapter } from '@salesforce/wire-service-jest-util';
+
+export const getTodo = createTestWireAdapter();
+```
````
is this a new file? if so, where would it live? under `__test__`? or?

jodarove replied:
it all depends on the jest configuration, but they usually are under `__mocks__` directory; in the sfdx, all the lightning modules are stubs via [the resolver](https://github.com/salesforce/sfdx-lwc-jest/blob/master/src/config.js)

yungcheng commented on README.md:
````diff
@@ -160,3 +170,36 @@ interface TestWireAdapter {
+// ... rest of the mocked module.
+```
+
+You will need to change that module mock and use one of the new `create*WireAdapterMock` methods introduced in version 3.x. Example:
````
```suggestion
You will need to change that module mock and use one of the new `create*WireAdapter` methods introduced in version 3.x. Example:
```

jodarove commented:
> As we are redesigning this API, I think it's a good time to address this specific issue #15. IMO, the fact that you can't target a specific adapter a huge flaw with this library. I don't think this proposal solves this specific issue.

i had it planned as a follow up pr but here it is: https://github.com/salesforce/wire-service-jest-util/pull/36/commits/c06305b05a57a5e67705973165a97297701dfe45

hlawrencesf commented:
I'm happy to open this as a separate issue, but one issue we've had on the original wire service mock is when one component has two different calls to the same wire adapter -- e.g. one object Info to some parent object, and one object info to some child object. In the past, it has not been possible to have different responses for the different inputs. If I'm interpreting this PR correctly, that will now be a supported scenario, right?

trevor-bliss commented:
@hlawrencesf Does the ability to filter on the `emit` API from Jose's branch he links to above solve your described issue?

https://github.com/salesforce/wire-service-jest-util/commit/c06305b05a57a5e67705973165a97297701dfe45#diff-1be3cf83f8983b51e6ae142eb4b35b177d064a089b92e91ff1a849db7d06637eR28

With this change you'd be able to filter what wire gets data emitted to based on a user defined filter function.

hlawrencesf commented:
I believe so. Reading through, I think something akin to 

```
testAdapterMock.emit(mockedValue1, (config) => config.p === 'first');

testAdapterMock.emit(mockedValue2, (config) => config.p === 'second');
```

would be supported, but I'm not familiar with this code base. If this example would work, then yep!

jodarove commented on README.md:
```diff
@@ -17,18 +17,21 @@ export default class MyComponent extends LightningElement {
 
 You'd like to test the component's handling of `@wire` data and errors. This test utility makes it trivial.
 
+Mock the `getTodo` wire adapter.
```
correct

trevor-bliss commented on src/observers/ApexWireAdapterObserver.js:
```diff
@@ -0,0 +1,41 @@
+        super.emit({ data: value, error: undefined }, filterFn);
+    }
+
+    error(body, status, statusText) {
```
I wonder if it makes sense to also filter for errors.

jodarove replied:
ill say yes! do you have something in mind?

- what about adding it at the end of the current `error`? (my first impression: I don't like it much)
- what about adding a new method `emitError({ body, status, statusText }, filterFn)` and maybe deprecate the current `error`?

trevor-bliss reviewed (commented):
This is great! One open question about filtering errors or not.

pmdartus commented on README.md:
```diff
@@ -144,14 +166,22 @@ interface FetchResponse {
+     *                 component with its associated config; if it returns true, the value will be
+     *                 emitted to that particular instance.
+     */
+    emit(value: object, filterFn?: (config) => boolean): void;
```
The testing API looks backward to me. As `emit` targets all the wire adapters, it is forcing us to pass a filter function to indicate which adapter you want to target. I am wondering if there is a better alternative. The core issue is that the test has to invoke `emit` and not the mocked wire adapter itself.

Instead of exposing the `emit` and `getLastConfig` outside the test wire adapter why not making the wire adapter self contained. 

```js
function imperative() {
    throw new Error('This adapter can\'t be directly invoked');
}

const adapter = createTestWireAdapter(imperative, function wired(config) {
    // Is invoked for each new config.

    // Invoke this.emit when you want to emit the value back to the component.
    this.emit({ sum: config.a + config.b });

    // [Apex and LDS adapters only] Invoke this.emitError when you want to emit an error.
    this.emitError({ body, status, statusText });

    setTimeout(() => {
        // The emit can be called any time (multiple time and async) if needed.
        this.emit({ delayed: 100 });
    }, 100);
})
```

pmdartus commented on src/observers/TestWireAdapterObserver.js:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    onConnect(instance) {
+        this._lastConfig = {};
```
Should `_lastConfig` still be `undefined` at this point ?

pmdartus commented on README.md:
```diff
@@ -129,6 +164,19 @@ interface ApexTestWireAdapter {
+     *                     component with its associated config; if it returns true, the error will be
+     *                     emitted to that particular instance.
+     */
+    emitError(
```
Is there any plan to depreate the `error` method or do you want to keep both `error` and `emitError` around?

jodarove replied:
I have mixed feelings about deprecating the existing `error` method. On one side, you could accomplish the same with `emitError`, on the other side, `error(body, status, statusText)` is currently used in tests, seems less verbose than `emitError` (IMO), and accommodate most of the use cases, also its maintenance is straightforward.

what's your opinion?

pmdartus replied:
Sounds good to me. Let's keep both around for backward compat purposes.

pmdartus commented on src/ObservableTestWireAdapter.js:
```diff
@@ -0,0 +1,39 @@
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+
+export class ObservableTestWireAdapter {
```
I don't understand the motivation behind this abstraction. Why do we need an `ObserbableWireAdapter` between the `Oberservers` and the adapter itself?

jodarove commented on src/index.js:
```diff
@@ -4,36 +4,125 @@
+import { TestWireAdapterObserver } from "./observers/TestWireAdapterObserver";
+import { registerAdapter } from "./utils";
+
+const WIRE_ADAPTER_MOCK_MARK = '$wire_adapter_mock$';
```
On SFDX, all modules exporting wire adapters are automatically mocked for customers. When this PR is merged, those automatic wire adapter mocks will be changed to a `TestWireAdapter` returned by this library, but the existing customer tests will still be using `register*WireAdapter`.

pmdartus commented on src/index.js:
```diff
@@ -4,36 +4,123 @@
+        testAdapter = fn;
+
+        fn.adapter = TestWireAdapter;
+        fn.emit = (...args) => TestWireAdapter.emit(...args);
+        if (TestWireAdapter.error) {
+            fn.error = (...args) => TestWireAdapter.error(...args);
+            fn.emitError = (...args) => TestWireAdapter.emitError(...args);
+        }
+        fn.getLastConfig = () => TestWireAdapter.getLastConfig();
```
Instead of handpicking the properties, what do you think about reassigning the whole prototype to `fn`?

```suggestion
        Object.setPrototypeOf(fn, TestWireAdapter);
```

The unintended side-effect is that it will expose all the "private" static properties like `_lastConfig` and `_wireInstances`. But I can live with this.

pmdartus commented on src/index.js:
```diff
@@ -4,36 +4,123 @@
+    if (typeof fn === "function") {
+        testAdapter = fn;
+
+        fn.adapter = TestWireAdapter;
```
Instead of associating `adapter` as an enumerable property it would make sense to `Object.defineProperty`:

```suggestion
        Object.defineProperty(fn, 'adapter', {
            value: TestWireAdapter,
        })
```

pmdartus reviewed (approved):
I have a couple of nitpicks, otherwise, this is a great PR. I would like is somebody else from the team to review this before merging. 

Now that I think about it, we should release a beta version before releasing the next major. I would like to try out the beta version on lwc-recipes, e-bikes and in Core. This way we can ensure the major version is bug-free.

ekashida reviewed (approved):
lgtm

jodarove commented:
I released a canary version based on this PR, in order to test these changes in core, lwc-recepies and e-bikes, noticing:

1. The package includes all tests, etc., but only the files under dist are required. Fixed in b790a85.

2. I got a SyntaxError on the dist because of `@babel/plugin-proposal-class-properties`. Fixed in 5ab726e.

3. There's was one missing detail that some core tests brought: `emit` on a disconnected component will push the value via wire to that component. It is [currently supported](https://github.com/salesforce/wire-service-jest-util/blob/master/src/adapters/generic.js#L14). Fixed in 964f060.

4. There's 1 test that loads the component using the wire after the `registerTestAdapter` call, this makes that the `wireAdaptersRegistryHack` being empty at the time of `registerTestAdapter`, and never spy on those adapters. In v2 this is not an issue because that use case was supported with the deprecated `wireService.register` [that was removed](https://github.com/salesforce/wire-service-jest-util/pull/35/files#diff-db2da36ff2913b723599796e99e6567f3ff9ee9b2b9adbbccb0ec863759aa969L11-L41). Migrating this test to the v3 API solves the issue.
